### PR TITLE
Consolidate and optimize `minimal_subalgebra_generators_with_relations`

### PR DESCRIPTION
**Pull request by joschmitt**
Documents and tests the (internal) function `minimal_subalgebra_generators_with_relations` and fixes some bugs found by testing...

Also renames the file `test/Rings/mpoly_affine_algebras.jl` to `test/Rings/mpoly-affine-algebra.jl`.

EDIT: This became a bit more now: I split up the function `subalgebra_membership_homogeneous` in two internal functions, so that one can reuse hard-to-compute data. This makes the logic of the code more complicated, but it reduced the runtime of `minimal_subalgebra_generators_with_relations` from ~4 seconds to ~1.2 seconds in a 'real life' example.

### CODE REVIEW

**Review by fingolfin — DISMISSED** [review]

**Comment by joschmitt** [comment]
To record the outcome of the discussion on last Wednesday:
* There are objections whether the function should be called `..._with_relations` because it does not return *all* relations between the polynomials.
* Moving around lists of polynomials which are 'interpreted' as a subalgebra should arguably not be the user-facing interface (I don't have an idea how it could be done better though).

File: src/Rings/mpoly-affine-algebras.jl [hunk 1]
```diff
@@ -1214,11 +1231,50 @@ julia> minimal_subalgebra_generators(V)
...
-  I = ideal(parent(V[1]), [ zero(parent(V[1])) ])
-  return _minimal_subalgebra_generators_with_relations(V, I, check = check)[1]
+  return minimal_subalgebra_generators_with_relations(V; check=check)[1]
 end
 
+@doc raw"""
+    minimal_subalgebra_generators_with_relations(V::Vector{T}; check::Bool = true) where T <: Union{MPolyRingElem, MPolyQuoRingElem}
+
+Given a vector `V` of homogeneous elements of a positively graded multivariate
+polynomial ring, or of a quotient of such a ring, return a minimal subset `W` of
```
Comment (fingolfin) [inline src/Rings/mpoly-affine-algebras.jl]: Minimal in which sense? By inclusion (i.e. no further element can be dropped from `W`), or by size (which imply the former but be stronger in general)?

File: src/exports.jl [hunk 2]
```diff
@@ -992,6 +992,7 @@ export minimal_nonfaces
 export minimal_normal_subgroups, has_minimal_normal_subgroups, set_minimal_normal_subgroups
 export minimal_primes
 export minimal_subalgebra_generators
+export minimal_subalgebra_generators_with_relations
```
Comment (fingolfin) [inline src/exports.jl]: I am (still) opposed to exporting this function under that name.

**Review by fingolfin — COMMENTED** [review]
I still think `minimal_subalgebra_generators_with_relations` is a horrible name, but this PR does not introduce it. We should of course allow to improve the existing code.

But I a concerned about exporting the name as we are then bound to provide it in all further OSCAR 1.x releases.

I'd be OK with merging this if that `export` goes away (and of course that would not preclude a future PR from exporting it, if circumstances change).

Perhaps I can discuss this with @fieker tomorrow.

File: src/Rings/mpoly-affine-algebras.jl (hunk 1 above)
Comment (joschmitt) [reply src/Rings/mpoly-affine-algebras.jl]: It's of minimal cardinality, I clarified the doc string.

File: src/exports.jl (hunk 2 above)
Comment (joschmitt) [reply src/exports.jl]: I removed the export.

**Comment by joschmitt** [comment]
Please review this. It contains bugfixes and performance improvements.

**Review by wdecker — COMMENTED** [review]
Why `Ideal(0)` with capital `I`? Also, should we replace `0` by the zero of the corresponding ring?

return `(true, ideal(0))` if the elements of `V` are algebraically independent over `K`.
return `(true, Ideal (0))` if the elements of `V` are algebraically independent over `K`.

**Review by wdecker — COMMENTED** [review]
We alread export `is_algebraically_independent_with_relations`. We must also export `minimal_subalgebra_generators_with_relations`.

**Comment by joschmitt** [comment]
> Why `Ideal(0)` with capital `I`? Also, should we replace `0` by the zero of the corresponding ring?

`Ideal (0)` is how the zero ideal would be printed. If we want to write the code to produce the zero ideal there, it would be something like `ideal([zero(parent(V[1]))])`.

**Comment by joschmitt** [comment]
> We alread export `is_algebraically_independent_with_relations`. We must also export `minimal_subalgebra_generators_with_relations`.

I think the function `is_algebraically_independent_with_relations` is actually a good argument to *not* export `minimal_subalgebra_generators_with_relations` under this name. The 'relations' returned by the latter function are not the relations of all generators, but only describe how the discarded generators can be represented in the chosen minimal set. I probably introduced this misleading name, but I can also not think of anything better apart from generic/vague things like `minimal_subalgebra_generators_with_data`.

**Comment by wdecker** [comment]
O.k., then leave it as it is.

**Comment by wdecker** [comment]
O.k., I misunderstood what the function is doing.

**Review by wdecker — APPROVED** [review]

**Pull request merged**